### PR TITLE
EOS-15950: Add benchmarking for ISA-L

### PR DESCRIPTION
**Pull request by sanjognaik**
- Added benchmarking for Intel(R) Intelligent Storage Acceleration Library (ISA-L)

Signed-off-by: Sanjog Vikram Naik <sanjog.naik@seagate.com>

### CODE REVIEW

**Comment by sanjognaik** [comment]
Below is unit benchmarking output on hardware for Galois and standalone ISA-L.
Galois:
`# ./scripts/m0 run-ub -t parity-math-ub`
`----- run_ub -t parity-math-ub -----`
`-- round 1parity-math-ub[.........]`
`               bench: [   iter]    min    max    avg   std   sec/op   op/sec block_size time_in_sec    bytes      KiB   BW_KiB     IOPS`
`set: parity-math-ub`
`         s 10/05/ 4K: [      1]   0.04   0.04   0.04  0.00% 3.846e-02/2.600e+01     00   0.04         0000     00     00     00`
`         m 20/06/ 4K: [      1]   0.21   0.21   0.21  0.00% 2.093e-01/4.778e+00     00   0.21         0000     00     00     00`
`         l 30/12/ 4K: [      1]   0.64   0.64   0.64  0.00% 6.402e-01/1.562e+00     00   0.64         0000     00     00     00`
`         s 10/05/32K: [      1]   0.26   0.26   0.26  0.00% 2.586e-01/3.866e+00     00   0.26         0000     00     00     00`
`         m 20/06/32K: [      1]   1.65   1.65   1.65  0.00% 1.648e+00/6.066e-01     00   1.65         0000     00     00     00`
`         l 30/12/32K: [      1]   5.05   5.05   5.05  0.00% 5.055e+00/1.978e-01     00   5.05         0000     00     00     00`
`        s  03/02/ 1M: [      1]   0.50   0.50   0.50  0.00% 5.038e-01/1.985e+00     00   0.50         0000     00     00     00`
`         m 06/03/ 1M: [      1]   2.25   2.25   2.25  0.00% 2.250e+00/4.444e-01     00   2.25         0000     00     00     00`
`         l 08/04/ 1M: [      1]   4.52   4.52   4.52  0.00% 4.519e+00/2.213e-01     00   4.52         0000     00     00     00
`

ISA-L:
`# ./scripts/m0 run-ub -t ec-isa-ub`
`----- run_ub -t ec-isa-ub -----`
`-- round 1ec-isa-ub[.........]`
`               bench: [   iter]    min    max    avg   std   sec/op   op/sec block_size time_in_sec    bytes      KiB   BW_KiB     IOPS`
`set: ec-isa-ub`
`         s 10/05/ 4K: [      1]   0.00   0.00   0.00  0.00% 9.190e-04/1.088e+03     00   0.00         0000     00     00     00`
`         m 20/06/ 4K: [      1]   0.00   0.00   0.00  0.00% 1.690e-03/5.917e+02     00   0.00         0000     00     00     00`
`         l 30/12/ 4K: [      1]   0.00   0.00   0.00  0.00% 2.667e-03/3.750e+02     00   0.00         0000     00     00     00`
`         s 10/05/32K: [      1]   0.01   0.01   0.01  0.00% 6.470e-03/1.546e+02     00   0.01         0000     00     00     00`
`         m 20/06/32K: [      1]   0.01   0.01   0.01  0.00% 1.290e-02/7.751e+01     00   0.01         0000     00     00     00`
`         l 30/12/32K: [      1]   0.02   0.02   0.02  0.00% 2.000e-02/5.000e+01     00   0.02         0000     00     00     00`
`        s  03/02/ 1M: [      1]   0.06   0.06   0.06  0.00% 6.348e-02/1.575e+01     00   0.06         0000     00     00     00`
`         m 06/03/ 1M: [      1]   0.12   0.12   0.12  0.00% 1.231e-01/8.125e+00     00   0.12         0000     00     00     00`
`         l 08/04/ 1M: [      1]   0.17   0.17   0.17  0.00% 1.669e-01/5.992e+00     00   0.17         0000     00     00     00`

File: sns/ut/isal_ub.c [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+/*
+ * Generate decode matrix from encode matrix and erasure list
+ *
+ */
+
+static int gf_gen_decode_matrix_simple(uint8_t * encode_matrix,
+                    uint8_t * decode_matrix,
+                    uint8_t * invert_matrix,
+                    uint8_t * temp_matrix,
+                    uint8_t * decode_index, uint8_t * frag_err_list, int nerrs, int k,
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:51]: I think we need to understand the current code and then re-write these two functions.
Or, we need to include the Intel copyright information/LICENSE.

What do you think?

Comment (sanjognaik) [reply sns/ut/isal_ub.c:51]: I have understood the code and now I am re-writing the function. Will commit the changes once done.

File: sns/ut/isal_ub.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    // Coefficient matrices
+    uint8_t *encode_matrix, *decode_matrix;
+    uint8_t *invert_matrix, *temp_matrix;
+    uint8_t *g_tbls;
+    uint8_t decode_index[MMAX];
+
+    m = k + p;
+
+    // Allocate coding matrices
+    encode_matrix = malloc(m * k);
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:124]: when re-writing the code, please use m0_alloc() instead.

File: sns/ut/isal_ub.c [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    decode_matrix = malloc(m * k);
+    invert_matrix = malloc(m * k);
+    temp_matrix = malloc(m * k);
+    g_tbls = malloc(k * p * 32);
+
+    if (encode_matrix == NULL || decode_matrix == NULL
+        || invert_matrix == NULL || temp_matrix == NULL || g_tbls == NULL) {
+        printf("Test failure! Error with malloc\n");
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:132]: Here, we need to:
```
m0_free(encode_matrix);
m0_free(decode_matrix);
m0_free(invert_matrix);
m0_free(temp_matrix);
m0_free(g_tbls);
```
Please note, m0_free(NULL) is OK.

File: sns/ut/isal_ub.c [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        return -1;
+    }
+
+    // Allocate the src & parity buffers
+    for (i = 0; i < m; i++) {
+        if (NULL == (frag_ptrs[i] = malloc(len))) {
+            printf("alloc error: Fail\n");
+            return -1;
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:140]: m0_free() all allocated memory.

Comment (huanghua78) [reply sns/ut/isal_ub.c:140]: And do cleanup elsewhere.

File: sns/ut/isal_ub.c [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    for (i = 0; i < p; i++)
+    {
+        uint32_t err_id = rand() % m;
+        frag_err_list[nerrs++] = err_id;
+        memset(frag_ptrs[err_id], 0xFF, len);
+    }
+
+    // Find a decode matrix to regenerate all erasures from remaining frags
+    gf_gen_decode_matrix_simple(encode_matrix, decode_matrix,
+                        invert_matrix, temp_matrix, decode_index,
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:174]: `invert_matrix` and `temp_matrix `  are only used inside this function.
We can allocate them inside the function `gf_gen_decode_matrix_simple()`.
It is important to reduce the scope of these variables.

File: sns/ut/isal_ub.c (hunk 2 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

File: sns/ut/isal_ub.c (hunk 3 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

File: sns/ut/isal_ub.c (hunk 4 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done

File: sns/ut/isal_ub.c (hunk 5 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

File: sns/ut/isal_ub.c [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+#define MMAX    255
+#define KMAX    255
+
+
+static int ub_init(const char *opts M0_UNUSED)
+{
+    time_t now;
+
+    time(&now);
+    srand(now);
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:44]: Motr has m0_time(), m0_rnd().

Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Used motr functions for the same.

File: Makefile.am [hunk 7]
```diff
@@ -753,7 +753,7 @@ utils/mkfs/m0mkfs.o: xcode/protocol_checksum.h
 sbin_PROGRAMS                  += utils/layout/m0layout
 utils_layout_m0layout_CPPFLAGS  = -DM0_TARGET='m0layout' $(AM_CPPFLAGS)
 utils_layout_m0layout_LDADD     = $(top_builddir)/motr/libmotr.la \
-                                  @MATH_LIBS@
+                                  @MATH_LIBS@ @ISAL_LIBS@
```
Comment (huanghua78) [inline Makefile.am:756]: Why is the `@ISAL_LIBS@` needed for this utility?
Why is it not needed for others?

I restored it and it builds successfully.

Comment (huanghua78) [reply Makefile.am:756]: The same for desim_ut_chs_test_LDADD

File: sns/ut/isal_ub.c [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+            err_list[i++] = err_id;
+        }
+    } while ( i < failed->b_nob );
+}
+
+static void encode_data(uint32_t k, uint32_t p, uint32_t len,
+                        uint8_t *encode_matrix, uint8_t *g_tbls,
+                        uint8_t **fragments)
+{
+    // Initialize g_tbls from encode matrix
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:76]: Please use /* */ style of comments. No C++ comments.

File: sns/ut/isal_ub.c [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    ec_init_tables(k, p, &encode_matrix[k * k], g_tbls);
+
+    // Generate EC parity blocks from sources
+    ec_encode_data(len, k, p, g_tbls, fragments, &fragments[k]);
+}
+
+static int decode_data(uint32_t k, uint32_t p, uint32_t len,
+                       uint8_t *encode_matrix, uint8_t *g_tbls,
+                       uint8_t **fragments,
+                       struct m0_buf failed, uint8_t **recover_outp)
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:86]: It's better to use "struct m0_buf *failed".

It is very confusing and error-prone to pass some struct, but the struct contains pointers.
The struct itself is copied to stack, but the memory pointed by the address is not copied.

File: ut/m0ub.c [hunk 10]
```diff
@@ -106,6 +107,7 @@ static void ub_add(const struct ub_args *args)
 	m0_ub_set_add(&m0_thread_ub);
 //	m0_ub_set_add(&m0_rpc_ub);
 //XXX_BE_DB	m0_ub_set_add(&m0_parity_math_ub);
```
Comment (huanghua78) [inline ut/m0ub.c:109]: `m0_parity_math_ub ` is not enabled.
How did you do the test?

File: Makefile.am (hunk 7 above)
Comment (huanghua78) [reply Makefile.am:756]: I have verified that without the "@ISAL_LIBS@", these two targets get built successfully.

Comment (sanjognaik) [reply Makefile.am]: Removed.

File: sns/ut/isal_ub.c (hunk 8 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

File: sns/ut/isal_ub.c (hunk 9 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

File: ut/m0ub.c (hunk 10 above)
Comment (sanjognaik) [reply ut/m0ub.c]: Enabled.

**Comment by sanjognaik** [comment]
Please refer the output and analyses of latest commit [here](https://jts.seagate.com/browse/EOS-15950?focusedCommentId=2014268&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-2014268).

File: sns/ut/parity_math_ut.c [hunk 11]
```diff
@@ -1434,48 +1434,111 @@ static void ub_large_32768(int iter)
...
         .us_init = ub_init,
         .us_fini = NULL,
         .us_run  = {
-                { .ub_name  = "s 10/05/ 4K",
+                { .ub_name  = "s 10/05/4K",
                   .ub_iter  = UB_ITER,
-                  .ub_round = ub_small_4096 },
+                  .ub_round = ub_small_4096,
+                  .ub_block_size = 4096,
+				  .ub_blocks_per_op = 15 },
```
Comment (huanghua78) [inline sns/ut/parity_math_ut.c:1475]: Indentation.

And similar ones in the following.

File: sns/ut/isal_ub.c [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+m0_time_t seed = 0;
+
+static int ub_init(const char *opts M0_UNUSED)
+{
+    seed = m0_time_now();
+
+    return 0;
+}
+
+static void unit_spoil( struct m0_buf *failed,
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:48]: No space is needed before the first param.
```
static void unit_spoil(struct m0_buf *failed,
```

File: sns/ut/isal_ub.c [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+                        const uint32_t data_count )
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:49]: And no space before the closing parenthesis.

File: sns/ut/isal_ub.c [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+            for (i = 0; i < k; i++)
+            {
+                s = 0;
+                for (j = 0; j < k; j++)
+                    s ^= gf_mul(invert_matrix[j * k + i],
+                            encode_matrix[k * err_list[p] + j]);
+                decode_matrix[k * p + i] = s;
+            }
+        }
+    }
```
Comment (huanghua78) [inline sns/ut/isal_ub.c:155]: Put the opening parenthesis at the same line.

**Review by huanghua78 — APPROVED** [review]
Looks fine.

Minor indentation and coding issues.

File: sns/ut/parity_math_ut.c (hunk 11 above)
Comment (sanjognaik) [reply sns/ut/parity_math_ut.c]: Done

File: sns/ut/isal_ub.c (hunk 12 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done

File: sns/ut/isal_ub.c (hunk 13 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done

File: sns/ut/isal_ub.c (hunk 14 above)
Comment (sanjognaik) [reply sns/ut/isal_ub.c]: Done.

**Comment by huanghua78** [comment]
Hello @sg-shankar, @nikitadanilov , @just-now , 
If you are interested in the detail implementation of the ISA UB and Galois UB, please take a look at this PR.

**Comment by sanjognaik** [comment]
This branch was created to get performance numbers of Intel ISA library and PR was created to get review comments only. It is not intended to land to main branch. So closing the PR.

**Pull request closed without merging**